### ISSUE TITLE
packetfabric_cloud_router - Intermittent 'Context Canceled' error on plan/apply

### ISSUE THREAD

**Issue by rparelius** [PacketFabric/terraform-provider-packetfabric#284]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform PacketFabric Provider Version

Terraform v1.1.6
on darwin_arm64
+ provider registry.terraform.io/hashicorp/aws v3.74.3
+ provider registry.terraform.io/packetfabric/packetfabric v0.4.2

### Affected Resource(s)

* packetfabric_cloud_router
* packetfabric_cloud_router_connection_google

### Terraform Configuration Files

Please include all Terraform configurations required to reproduce the bug. Bug reports without a functional reproduction may be closed without investigation.

**main.tf**
```hcl
terraform {
  required_providers {
    packetfabric = {
      source  = "PacketFabric/packetfabric"
      version = ">= 0.4.2"
    }
  }
}

provider "packetfabric" {
  host = "https://api.packetfabric.com"
  token = var.pf_api_key
  alias = "pf"
}

module "packetfabric-cloud-router" {
  source = "./modules/packetfabric"
  for_each = var.cloud_routers
  aws_account_id = var.aws_account_id
  pf_account_uuid = "redacted"
  name = each.key 
  asn = each.value.asn 
  capacity = each.value.capacity 
  regions = each.value.regions
  aws_connections = var.aws_connections[each.key]
  aws_bgp_sessions = var.aws_bgp_sessions[each.key]
  aws_inbound = var.aws_inbound[regex("\\w+-\\w+$", each.key)]
  aws_outbound = var.aws_outbound[regex("\\w+-\\w+$", each.key)]
  gcp_connections = var.gcp_connections[each.key]
  gcp_bgp_sessions = var.gcp_bgp_sessions[each.key]
  gcp_inbound = var.gcp_inbound[regex("\\w+-\\w+$", each.key)]
  gcp_outbound = var.gcp_outbound[regex("\\w+-\\w+$", each.key)]
  providers = {
    packetfabric = packetfabric.pf
  }
}
```

**modules/packetfabric/cr.tf**
```hcl
variable "pf_account_uuid" {
  type = string
}

variable "aws_account_id" {
  type = string
}

variable "asn" {
  type = number 
}

variable "name" {
  type = string
}

variable "capacity" {
  type = string
}

variable "regions" {
  type = list(string)
}

variable "aws_connections" {
  type = map(any)
}

variable "gcp_connections" {
  type = map(any)
}

variable "gcp_bgp_sessions" {
  type = map(any)
}

variable "aws_bgp_sessions" {
  type = map(any)
}

variable "gcp_inbound" {
  type = list(any)
}

variable "gcp_outbound" {
  type = list(any)
}

variable "aws_inbound" {
  type = list(any)
}

variable "aws_outbound" {
  type = list(any)
}

terraform {
  required_providers {
    packetfabric = {
      source  = "PacketFabric/packetfabric"
      version = ">= 0.4.2"
    }
  }
}

resource "packetfabric_cloud_router" "cr" {
  provider     = packetfabric
  account_uuid = var.pf_account_uuid
  asn          = var.asn
  name         = var.name
  capacity     = var.capacity
  regions      = var.regions
  timeouts {
    create = "20m"
  }
}

resource "packetfabric_cloud_router_connection_aws" "aws" {
  for_each = var.aws_connections
  provider = packetfabric
  account_uuid = var.pf_account_uuid
  aws_account_id = var.aws_account_id
  circuit_id = packetfabric_cloud_router.cr.id
  description = "aws-${each.key}"
  pop = each.value.pop
  speed = each.value.speed
  zone = each.value.zone
  timeouts {
    create = "20m"
  }
}

resource "packetfabric_cloud_router_connection_google" "gcp" {
  for_each = var.gcp_connections
  provider                    = packetfabric
  circuit_id                  = packetfabric_cloud_router.cr.id
  account_uuid                = var.pf_account_uuid
  google_pairing_key          = each.value.pairing_key
  google_vlan_attachment_name = each.value.vlan_attachment_name
  pop                         = each.value.pop
  speed                       = each.value.speed
  description                 = "gcp-${each.key}"
  timeouts {
    create = "20m"
  }
}


resource "packetfabric_cloud_router_bgp_session" "gcp" {
  for_each = var.gcp_bgp_sessions
  provider       = packetfabric
  circuit_id     = packetfabric_cloud_router.cr.id
  connection_id  = packetfabric_cloud_router_connection_google.gcp[each.key].id
  address_family = "v4"
  remote_asn     = each.value.asn
  remote_address = each.value.remote_address
  l3_address     = each.value.l3_address
  bfd_interval   = 1000
  bfd_multiplier = 5
  multihop_ttl   = 1
  disabled       = each.value.disabled

  dynamic "prefixes" {
    for_each   = { for i,v in var.gcp_inbound : i => v if contains(v.env, each.key)}
    content {
      prefix     = prefixes.value.prefix
      type       = "in"
      match_type = prefixes.value.match_type
      order      = prefixes.key
    }
  }

  dynamic "prefixes" {
    for_each   = { for i,v in var.gcp_outbound : i => v }
    content {
      prefix     = prefixes.value.prefix
      type       = "out"
      match_type = prefixes.value.match_type
      med        = lookup(prefixes.value, "med", 0)
      order      = prefixes.key
    }
  }
}


resource "packetfabric_cloud_router_bgp_session" "aws" {
  for_each = var.aws_bgp_sessions
  provider       = packetfabric
  circuit_id     = packetfabric_cloud_router.cr.id
  connection_id  = packetfabric_cloud_router_connection_aws.aws[each.key].id
  address_family = "v4"
  remote_asn     = each.value.asn
  remote_address = each.value.remote_address
  l3_address     = each.value.l3_address
  bfd_interval   = 300
  bfd_multiplier = 3
  multihop_ttl   = 1
  md5            = each.value.md5
  disabled       = each.value.disabled

  dynamic "prefixes" {
    for_each   = { for i,v in var.aws_inbound : i => v }
    content {
      prefix     = prefixes.value.prefix
      type       = "in"
      match_type = prefixes.value.match_type
      order      = prefixes.key
    }
  }

  dynamic "prefixes" {
    for_each   = { for i,v in var.aws_outbound : i => v }
    content {
      prefix     = prefixes.value.prefix
      type       = "out"
      match_type = prefixes.value.match_type
      order      = prefixes.key
    }
  }
}

```

**tfvars.json**
```json
{
    "aws_account_id": "redacted",
    "cloud_routers": {
        "dev-us-west": {
            "asn": 4556,
            "capacity": "100Gbps",
            "regions": ["US"]
        },
        "dev-us-east": {
            "asn": 4556,
            "capacity": "100Gbps",
            "regions": ["US"]
        }
    },
    "aws_connections": {
        "dev-us-west": {
            "dxg": {
                "speed": "50Mbps",
                "pop": "PDX2",
                "zone": "A"
            }
        },
        "dev-us-east": {
            "dxg": {
                "speed": "50Mbps",
                "pop": "WDC2",
                "zone": "E"
            }
        }
    },
    "gcp_connections": {
        "dev-us-west": {
            "dev": {
                "pairing_key": "redacted",
                "vlan_attachment_name": "vlan-dev-pf",
                "speed": "50Mbps",
                "pop": "SEA3"
            }
        },
        "dev-us-east": {
            "dev": {
                "pairing_key": "redacted",
                "vlan_attachment_name": "vlan-dev-pf",
                "speed": "50Mbps",
                "pop": "WDC1"
            }
        }
    },
    "gcp_bgp_sessions": {
        "dev-us-west": {
        },
        "dev-us-east": {
        }
    },
    "aws_bgp_sessions": {
        "dev-us-west": {
        },
        "dev-us-east": {
        }
    },
    "gcp_inbound": {
        "us-west": [
            { "prefix": "10.0.0.0/8", "match_type": "orlonger", "env": ["dev", "stage", "prod"] },
            { "prefix": "35.199.192.0/19", "match_type": "exact", "env": ["prod"] },
            { "prefix": "199.36.153.4/30", "match_type": "exact", "env": ["prod"] },
            { "prefix": "199.36.153.8/30", "match_type": "exact", "env": ["prod"] }
        ],
        "us-east": [
            { "prefix": "10.0.0.0/8", "match_type": "orlonger", "env": ["dev", "stage", "prod"] },
            { "prefix": "35.199.192.0/19", "match_type": "exact", "env": ["prod"] },
            { "prefix": "199.36.153.4/30", "match_type": "exact", "env": ["prod"] },
            { "prefix": "199.36.153.8/30", "match_type": "exact", "env": ["prod"] }
        ]
    },
    "gcp_outbound": {
        "us-west": [
            { "prefix": "1.1.1.1/32", "match_type": "exact" }
        ],
        "us-east": [
            { "prefix": "1.1.1.1/32", "match_type": "exact" }
        ]
    },
    "aws_inbound": {
        "us-west": [
            { "prefix": "10.0.0.0/8", "match_type": "orlonger" }
        ],
        "us-east": [
            { "prefix": "10.0.0.0/8", "match_type": "orlonger" }
        ]
    },
    "aws_outbound": {
        "us-west": [
            { "prefix": "1.1.1.1/32", "match_type": "exact" }
        ],
        "us-east": [
            { "prefix": "1.1.1.1/32", "match_type": "exact" }
        ]
    }
}
```
### Debug Output

https://gist.github.com/rparelius/53432d50911686438a071d0e0011d336

### Expected Behavior

Terraform should successfully generate a plan.

### Actual Behavior

terraform plan/apply errors out with `context canceled` error

### Steps to Reproduce

The resources must exist before it errors out during planning. Once created, if I try to destroy the resources, it errors out when executing the plan. 

```bash
terraform plan --var-file=tfvars.json
```

The error happens intermittently, just continue to `terraform plan` and it eventually happens. for me it happens about 25% of the time.

**Comment by rjouhann** [comment]
Hello @rparelius,

I noticed you are using Terraform `v1.1.6`, could you maybe try to use the [latest version](https://developer.hashicorp.com/terraform/downloads) `v1.3.5`?

I also found this post on [StackOverflow](https://stackoverflow.com/questions/72661399/intermittent-context-canceled-error-on-terraform-plan) and they suggest to increased the Compute resources allocated to run terraform. Could you look into this?

Thanks & Regards,
Romain

**Comment by rparelius** [comment]
> Hello @rparelius,
> 
> I noticed you are using Terraform `v1.1.6`, could you maybe try to use the [latest version](https://developer.hashicorp.com/terraform/downloads) `v1.3.5`?
> 
> I also found this post on [StackOverflow](https://stackoverflow.com/questions/72661399/intermittent-context-canceled-error-on-terraform-plan) and they suggest to increased the Compute resources allocated to run terraform. Could you look into this?
> 
> Thanks & Regards, Romain

Hi @rjouhann, I am getting this error even on my Macbook Pro with both v1.1.6 and v.1.3.5
sometimes the plan succeeds, sometimes it just fails:


```bash
~/Development/tf-test ./terraform-1-3-5 plan --var-file=tfvars.json
module.packetfabric-cloud-router["dev-us-west"].packetfabric_cloud_router.cr: Refreshing state... [id=PF-L3-CUST-1753834]
module.packetfabric-cloud-router["dev-us-east"].packetfabric_cloud_router.cr: Refreshing state... [id=PF-L3-CUST-1753833]
╷
│ Error: Get "https://api.packetfabric.com/v2/services/cloud-routers/PF-L3-CUST-1753834": context canceled
│ 
│   with module.packetfabric-cloud-router["dev-us-west"].packetfabric_cloud_router.cr,
│   on modules/packetfabric/main.tf line 10, in resource "packetfabric_cloud_router" "cr":
│   10: resource "packetfabric_cloud_router" "cr" {
│ 
╵
~/Development/tf-test ./terraform-1-3-5 plan --var-file=tfvars.json
module.packetfabric-cloud-router["dev-us-east"].packetfabric_cloud_router.cr: Refreshing state... [id=PF-L3-CUST-1753833]
module.packetfabric-cloud-router["dev-us-west"].packetfabric_cloud_router.cr: Refreshing state... [id=PF-L3-CUST-1753834]
module.packetfabric-cloud-router["dev-us-west"].packetfabric_cloud_router_connection_google.gcp["dev"]: Refreshing state... [id=PF-L3-CON-1753846]
module.packetfabric-cloud-router["dev-us-east"].packetfabric_cloud_router_connection_aws.aws["dxg"]: Refreshing state... [id=PF-L3-CON-1753838]
module.packetfabric-cloud-router["dev-us-east"].packetfabric_cloud_router_connection_google.gcp["dev"]: Refreshing state... [id=PF-L3-CON-1753845]
module.packetfabric-cloud-router["dev-us-west"].packetfabric_cloud_router_connection_aws.aws["dxg"]: Refreshing state... [id=PF-L3-CON-1753852]

No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.
~/Development/tf-test ./terraform-1-3-5 plan --var-file=tfvars.json
module.packetfabric-cloud-router["dev-us-east"].packetfabric_cloud_router.cr: Refreshing state... [id=PF-L3-CUST-1753833]
module.packetfabric-cloud-router["dev-us-west"].packetfabric_cloud_router.cr: Refreshing state... [id=PF-L3-CUST-1753834]
╷
│ Error: Get "https://api.packetfabric.com/v2/services/cloud-routers/PF-L3-CUST-1753833": context canceled
│ 
│   with module.packetfabric-cloud-router["dev-us-east"].packetfabric_cloud_router.cr,
│   on modules/packetfabric/main.tf line 10, in resource "packetfabric_cloud_router" "cr":
│   10: resource "packetfabric_cloud_router" "cr" {
│ 
╵
~/Development/tf-test 
```

**Comment by rjouhann** [comment]
I was able to repro the issue. I made your sample code a bit shorter to only include GCP.

Debug Log for my repro: https://gist.github.com/rjouhann/1169467fb58aa64231bb04c0c5574ef9

Ran:
```
terraform apply --var-file=tfvars.json (OK)
terraform plan --var-file=tfvars.json (OK)
terraform plan --var-file=tfvars.json (OK)
terraform plan --var-file=tfvars.json (NOK)
```

**main.tf**
```tf
terraform {
  required_providers {
    packetfabric = {
      source  = "PacketFabric/packetfabric"
      version = ">= 0.4.2"
    }
  }
}

provider "packetfabric" {
  # host = "https://api.packetfabric.com" # set by default in the provider
  # token = var.pf_api_key # set via an env variable
  alias = "pf"
}

module "packetfabric-cloud-router" {
  source = "./modules/packetfabric"
  for_each = var.cloud_routers
  # pf_account_uuid = "redacted" # set via an env variable
  name = each.key
  asn = each.value.asn
  capacity = each.value.capacity
  regions = each.value.regions
  gcp_connections = var.gcp_connections[each.key]
  gcp_bgp_sessions = var.gcp_bgp_sessions[each.key]
  gcp_inbound = var.gcp_inbound[regex("\\w+-\\w+$", each.key)]
  gcp_outbound = var.gcp_outbound[regex("\\w+-\\w+$", each.key)]
  providers = {
    packetfabric = packetfabric.pf
  }
}

# declare variables in the root module
variable "cloud_routers" {
  type = map(any)
}

variable "gcp_connections" {
  type = map(any)
}

variable "gcp_bgp_sessions" {
  type = map(any)
}

variable "gcp_inbound" {
  type = map(any)
  # type = list(any)
}

variable "gcp_outbound" {
  type = map(any)
  # type = list(any)
}
```

**modules/packetfabric/cr.tf**
```tf
variable "asn" {
  type = number 
}

variable "name" {
  type = string
}

variable "capacity" {
  type = string
}

variable "regions" {
  type = list(string)
}

variable "gcp_connections" {
  type = map(any)
}

variable "gcp_bgp_sessions" {
  type = map(any)
}

variable "gcp_inbound" {
  type = list(any)
}

variable "gcp_outbound" {
  type = list(any)
}

terraform {
  required_providers {
    packetfabric = {
      source  = "PacketFabric/packetfabric"
      version = ">= 0.4.2"
    }
  }
}

resource "packetfabric_cloud_router" "cr" {
  provider     = packetfabric
  asn          = var.asn
  name         = var.name
  capacity     = var.capacity
  regions      = var.regions
  timeouts {
    create = "20m"
  }
}

resource "packetfabric_cloud_router_connection_google" "gcp" {
  for_each = var.gcp_connections
  provider                    = packetfabric
  circuit_id                  = packetfabric_cloud_router.cr.id
  google_pairing_key          = each.value.pairing_key
  google_vlan_attachment_name = each.value.vlan_attachment_name
  pop                         = each.value.pop
  speed                       = each.value.speed
  description                 = "gcp-${each.key}"
  timeouts {
    create = "20m"
  }
}


resource "packetfabric_cloud_router_bgp_session" "gcp" {
  for_each = var.gcp_bgp_sessions
  provider       = packetfabric
  circuit_id     = packetfabric_cloud_router.cr.id
  connection_id  = packetfabric_cloud_router_connection_google.gcp[each.key].id
  address_family = "v4"
  remote_asn     = each.value.asn
  remote_address = each.value.remote_address
  l3_address     = each.value.l3_address
  bfd_interval   = 1000
  bfd_multiplier = 5
  multihop_ttl   = 1
  disabled       = each.value.disabled

  dynamic "prefixes" {
    for_each   = { for i,v in var.gcp_inbound : i => v if contains(v.env, each.key)}
    content {
      prefix     = prefixes.value.prefix
      type       = "in"
      match_type = prefixes.value.match_type
      order      = prefixes.key
    }
  }

  dynamic "prefixes" {
    for_each   = { for i,v in var.gcp_outbound : i => v }
    content {
      prefix     = prefixes.value.prefix
      type       = "out"
      match_type = prefixes.value.match_type
      med        = lookup(prefixes.value, "med", 0)
      order      = prefixes.key
    }
  }
}
```

**tfvars.json**
```json
{
    "cloud_routers": {
        "dev-us-west": {
            "asn": 4556,
            "capacity": "100Gbps",
            "regions": ["US"]
        },
        "dev-us-east": {
            "asn": 4556,
            "capacity": "100Gbps",
            "regions": ["US"]
        }
    },
    "gcp_connections": {
        "dev-us-west": {
            "dev": {
                "pairing_key": "redacted",
                "vlan_attachment_name": "vlan-dev-pf",
                "speed": "50Mbps",
                "pop": "SEA3"
            }
        },
        "dev-us-east": {
            "dev": {
                "pairing_key": "redacted",
                "vlan_attachment_name": "vlan-dev-pf",
                "speed": "50Mbps",
                "pop": "WDC1"
            }
        }
    },
    "gcp_bgp_sessions": {
        "dev-us-west": {
        },
        "dev-us-east": {
        }
    },
    "gcp_inbound": {
        "us-west": [
            { "prefix": "10.0.0.0/8", "match_type": "orlonger", "env": ["dev", "stage", "prod"] },
            { "prefix": "35.199.192.0/19", "match_type": "exact", "env": ["prod"] },
            { "prefix": "199.36.153.4/30", "match_type": "exact", "env": ["prod"] },
            { "prefix": "199.36.153.8/30", "match_type": "exact", "env": ["prod"] }
        ],
        "us-east": [
            { "prefix": "10.0.0.0/8", "match_type": "orlonger", "env": ["dev", "stage", "prod"] },
            { "prefix": "35.199.192.0/19", "match_type": "exact", "env": ["prod"] },
            { "prefix": "199.36.153.4/30", "match_type": "exact", "env": ["prod"] },
            { "prefix": "199.36.153.8/30", "match_type": "exact", "env": ["prod"] }
        ]
    },
    "gcp_outbound": {
        "us-west": [
            { "prefix": "1.1.1.1/32", "match_type": "exact" }
        ],
        "us-east": [
            { "prefix": "1.1.1.1/32", "match_type": "exact" }
        ]
    }
}
```

**Comment by rjouhann** [comment]
@rparelius we are still looking into the issue but I think I have a workaround for you in the meantime.

When adding [-parallelism=1](https://developer.hashicorp.com/terraform/cli/commands/plan#parallelism-n) in `terraform plan --var-file=tfvars.json` command, the _Context Canceled error_ doesn't occur.

**Comment by rparelius** [comment]
Hi @rjouhann , just wondering if this could be due to the `sendRequest` function reusing the context for subsequent calls https://github.com/PacketFabric/terraform-provider-packetfabric/blob/main/internal/packetfabric/client.go#L102-L108

```go
if c.Ctx == nil {
	c.Ctx = context.Background()
}
...
req, _ = http.NewRequestWithContext(c.Ctx, method, formatedURL, nil)
```

could a new context be created for each request?

**Comment by rjouhann** [comment]
This bug will be fixed as part of `v0.5.1` scheduled to go out today.
